src/Tools/dotnet-monitor/ParameterCapturing/EventParameterCapturingPipeline.cs

```diff
@@ -27,6 +30,8 @@ internal sealed class ServiceStateUpdateArgs
 
     internal sealed class EventParameterCapturingPipeline : EventSourcePipeline<EventParameterCapturingPipelineSettings>
     {
+        private readonly EventParameterCapturingPipelineCache _cache = new();
```

Comment:
I don't think a cache is the right concept here -- it has a few undesired behaviors such as:
- Unbounded memory growth when dealing with long running parameter capturing sessions
- Possible data inconsistency, e.g. let's say the sender is buggy and somehow we receive an `CapturedParameter` event with a requestId that corresponds to a previously seen `CapturedParameter/Stop` event. We'd be updated an already stopped request.


My suggestion:
The cache should become more of a builder (`CapturedParametersBuilder` or something similar) that lets you:
- Start building a new set of captured parameters for a specific `requestId`.
- Add parameters to an existing set of captured parameters via requestId`.
- Finalize a set of captured parameters, returning the fully built set and removing the builders knowledge of it to help bound memory usage.


It's pretty similar to what you have now, just with the notion of finalizing sets of parameters.

Reply:
Thinking about this further, the way you currently have `AsyncParameterCapturingEventSource` written there will only ever be 1 set of captured parameters being written to the event source at a time (e.g. no interweaving of different sets of parameters). This simplifies the requirements here quite a bit as we don't need to store multiple requests at a time. 

We really only need to store a single `CapturedParameters` and Guid for the active `requestId` at any given time.  If we ever receive `CapturedParameter` for a non-active `requestId` we can just throw away that event data, and if we ever receive a `CapturedParameter/Start` for a different `requestId` while another one is still active we can throw out the current request being built.

With all that said, I'm fine either way here and my earlier suggestions for `CapturedParametersBuilder` would be more versatile with smaller changes to your PR so I'll defer to you on which approach you want to take.

Reply:
I tried to follow the same pattern as in the exceptions operations: the factory is transient and the store is scoped to the request (see `ServiceCollectionExtensions.ConfigureParameterCapturing`), so because the pipeline's lifetime is tied to the `CaptureParametersOperation`, the cache will never grow unbounded and it will get destroyed once the request completes.

Reply:
> so because the pipeline's lifetime is tied to the CaptureParametersOperation, the cache will never grow unbounded and it will get destroyed once the request completes.

It will grow unbounded though given a parameter capturing session with no capture limit, as each time a method with probes in encountered we grow our cache here with no way of freeing that memory until the session is done. The difference between this and the exceptions operation is the exception operation needs to maintain a cache for type names (which is inherently bounded by the number of types in the target application). We don't need to maintain data about prior parameter sets so we shouldn't continue to store that data after it's useful.

Reply:
Sure, I can implement your proposal, but we'll still grow the memory in an unbounded manner on a higher layer, in `ParameterCapturingStore`

Reply:
>but we'll still grow the memory in an unbounded manner on a higher layer, in ParameterCapturingStore

And that's also a problem :) `ParameterCapturingStore` should be bounded (I'd argue we shouldn't have it in the first place but I'll leave a separate comment for that).